packages/devtools_app/lib/src/shared/utils/profiler_utils.dart

```diff
@@ -177,18 +182,28 @@ class BottomUpTransformer<T extends ProfilableDataMixin<T>> {
   /// Returns the roots for a bottom up representation of a
   /// [ProfilableDataMixin] node.
   ///
+  /// The [stopwatch] is used to chunk up work and try to avoid dropping frames,
+  /// and doesn't need to be passed in from the outside.
+  ///
   /// Each root is a leaf from the original [ProfilableDataMixin] tree, and its
   /// children will be the reverse stack of the original profile sample. The
   /// stack returned will not be merged to combine common roots, and the sample
   /// counts will not reflect the bottom up sample counts. These steps will
   /// occur later in the bottom-up conversion process.
   @visibleForTesting
-  List<T> generateBottomUpRoots({
+  Future<List<T>> generateBottomUpRoots({
     required T node,
-    required T? currentBottomUpRoot,
+    required T? parent,
     required List<T> bottomUpRoots,
     bool skipRoot = false,
-  }) {
+    Stopwatch? stopwatch,
+  }) async {
+    stopwatch ??= Stopwatch()..start();
+    if (stopwatch.elapsedMilliseconds > frameBudgetMs * 0.5) {
+      await delayToReleaseUiThread(micros: 5000);
+      stopwatch.reset();
+    }
```

Comment:
can we put this in a helper somewhere? we are using this logic several times now and this would be generally useful to have some `maybeReleaseUiThread` helper

Reply:
Unfortunately not much can be extracted... I could make an extension on Stopwatch to replace the if expression with something like `stopwatch.exceedsFrameBudget`.

It is kind of hard to explain, but the stopwatch must be managed by the caller here, because otherwise it will start running before we actually resume executing this chunk of work (as soon as the future completes, but before this async task is actually picked back up).

Reply:
We also want to continue synchronously which would mean having to deal with FutureOr instances if we shelled out to a helper and that would not end up looking any cleaner really.